snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,421 @@
+
+* **Promote snapshot to PV:**
+
+    1. The user creates a persistent volume claim referencing the snapshot object in the annotation. The PVC must
+       belong to a `StorageClass` using the external volume snapshot provisioner.
+
+    2. The provisioner will use the `VolumeSnapshotData` object to create a persistent volume using the corresponding
+       volume snapshot plugin.
+
+    3. The PVC is bound to the newly created PV containing the data from the snapshot.
+
+
+The snapshot operation is a no-op for volume plugins that do not support snapshots via an API call (i.e. non-cloud storage).
+
+
+### API
+
+* The `VolumeSnapshot` object
+
+```
+// The volume snapshot object accessible to the user. Upon succesful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotData through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta
+	Metadata        metav1.ObjectMeta
+
+	// Spec represents the desired state of the snapshot
+	Spec VolumeSnapshotSpec
+
+	// Status represents the latest observer state of the snapshot
+	Status VolumeSnapshotStatus
+}
+
+// The desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+	// PersistentVolumeClaimName is the name of the PVC being snapshotted
+	PersistentVolumeClaimName string
+
+	// SnapshotDataName binds the VolumeSnapshot object with the VolumeSnapshotData
+	SnapshotDataName string
+}
+
+```
+
+* The `VolumeSnapshotData` object
+
+```
+// VolumeSnapshotData represents the actual "on-disk" snapshot object
+type VolumeSnapshotData struct {
+	metav1.TypeMeta
+	Metadata metav1.ObjectMeta
+
+	// Spec represents the desired state of the snapshot
+	Spec VolumeSnapshotDataSpec
+
+	// Status represents the latest observed state of the snapshot
+	Status VolumeSnapshotDataStatus
+}
+
+// The desired state of the volume snapshot
+type VolumeSnapshotDataSpec struct {
+	// Source represents the location and type of the volume snapshot
+	VolumeSnapshotDataSource
+
+	// VolumeSnapshotRef is part of bi-directional binding between VolumeSnapshot
+	// and VolumeSnapshotData
+	VolumeSnapshotRef *core_v1.ObjectReference
+
+	// PersistentVolumeRef represents the PersistentVolume that the snapshot has been
+	// taken from
+	PersistentVolumeRef *core_v1.ObjectReference
+}
+
+// Represents the actual location and type of the snapshot. Only one of its members may be specified.
+type VolumeSnapshotDataSource struct {
+    // Per-plugin specific spec
+}
+```
+
+* Event loop in controller
+
+The volume snapshot controller maintains two data structures (ActualStateOfWorld and DesiredStateOfWorld) and
+periodically reconciles the two. The data structures are being update by the API sever event handlers.
+
+* If a new `VolumeSnapshot` is added, the *add* handler adds it to the DesiredStateOfWorld (DSW)
+
+* If a `VolumeSnapshot` is deleted, the *delete* handler removes it from the DSW.
+
+* Reconciliation loop in the controller
```

Comment:
There is some rudimentary state check in the reconciler too. It checks whether the objects present in both states of world has all the required attributes set correctly and fixes it eventually. It's there primarily to correct eventual failed API server operations.